index.js

```diff
@@ -5,46 +5,69 @@ var DEFAULT_FILTER = /^([^\.].*)\.js(on)?$/;
-  var filter = options.filter === undefined ? DEFAULT_FILTER : options.filter;
-  var modules = {};
-  var recursive = options.recursive === undefined ? DEFAULT_RECURSIVE : options.recursive;
-  var resolve = options.resolve || identity;
-  var map = options.map || identity;
-
-  function excludeDirectory(dirname) {
-    return !recursive ||
-      (excludeDirs && dirname.match(excludeDirs));
-  }
-
-  var files = fs.readdirSync(dirname);
-
-  files.forEach(function (file) {
-    var filepath = dirname + '/' + file;
-    if (fs.statSync(filepath).isDirectory()) {
-
-      if (excludeDirectory(file)) return;
-
-      modules[map(file, filepath)] = requireAll({
-        dirname: filepath,
-        filter: filter,
-        excludeDirs: excludeDirs,
-        map: map,
-        resolve: resolve
-      });
-
-    } else {
-      var match = file.match(filter);
-      if (!match) return;
-
-      modules[map(match[1], filepath)] = resolve(require(filepath));
+    var dirname = typeof options === 'string' ? options : options.dirname;
+    var excludeDirs = options.excludeDirs === undefined ? DEFAULT_EXCLUDE_DIR : options.excludeDirs;
+    var filter = options.filter === undefined ? DEFAULT_FILTER : options.filter;
+    var modules = {};
+    var recursive = options.recursive === undefined ? DEFAULT_RECURSIVE : options.recursive;
+    var resolve = options.resolve || identity;
+    var map = options.map || identity;
+
+    function excludeDirectory(dirname) {
+        return !recursive ||
+            (excludeDirs && dirname.match(excludeDirs));
     }
-  });
 
-  return modules;
+    var files = fs.readdirSync(dirname);
+
+    files.forEach(function (file) {
+        var filepath = dirname + '/' + file;
+        if (fs.statSync(filepath).isDirectory()) {
+
+            if (excludeDirectory(file)) return;
+
+            modules[map(file, filepath)] = requireAll({
+                dirname: filepath,
+                filter: filter,
+                excludeDirs: excludeDirs,
+                map: map,
+                resolve: resolve
+            });
+
+        } else {
+            if (filter instanceof RegExp) {
+                var match = file.match(filter);
+                if (!match) return;
+
+                modules[map(match[1], filepath)] = resolve(require(filepath));
+            } else {
+                var match = filter instanceof Function ? filterOutFuncHandler(file, filter) :
+                    filter instanceof Array ? filterOutArrayHandler(file, filter) : defaultError();
+                if (match) return;
+                var fileName = file.replace(/\.[^/.]+$/, "")
+                modules[map(fileName, filepath)] = resolve(require(filepath));
+            }
+        }
+    });
+
+    return modules;
 };
 
 function identity(val) {
-  return val;
+    return val;
+}
+
+function filterOutFuncHandler(file, filter) {
+    var result = filter(file);
+    if (typeof result === "boolean") return result;
```

Comment:
Why so strict? In fact, this is a lame interface compared to the regular expression one, even though functions are more flexible. The regular expression one allows you to use capture groups to trim down what `map` gets. Can you make it just let through returning a string as being `true` with a different name to use?